### PR TITLE
fix: fix out of order boundary check [DET-4422]

### PR DESCRIPTION
**Pull request by hamidzr**
## Description

from 
`{"error":{"code":13,"reason":"Internal","error":"reflect.Value.Slice: slice index out of bounds"}}`
to 
`{"error":{"code":3,"reason":"InvalidArgument","error":"offset out of bounds"}}`

## Checklist

- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: master/internal/api.go [hunk 1]
```diff
@@ -46,10 +46,14 @@ func (a *apiServer) paginate(p **apiv1.Pagination, values interface{}, offset, l
 		EndIndex:   endIndex,
 		Total:      total,
 	}
-	rv.Elem().Set(rv.Elem().Slice(int(startIndex), int(endIndex)))
-	return grpc.ValidateRequest(
+	err := grpc.ValidateRequest(
 		func() (bool, string) { return 0 <= startIndex && startIndex <= total, "offset out of bounds" },
```
Comment (mackrorysd) [inline master/internal/api.go:50]: Is startIndex == total valid? Seems to me that would be out of bounds, and we should check startIndex < total.

**Review by mackrorysd — COMMENTED** [review]
Change looks good. 1 nitpick on an extant issue.

File: master/internal/api.go (hunk 1 above)
Comment (hamidzr) [reply master/internal/api.go:50]: that stood out to me too but if I'm not mistaken it is. experiment ids start at 1.

{"experiments":[],"pagination":{"offset":67,"limit":0,"startIndex":67,"endIndex":67,"total":67}}

**Review by mackrorysd — APPROVED** [review]

**Comment by mackrorysd** [comment]
Go ahead and merge, if ready. My next question was going to be, why is startIndex == 0 valid, but I guess that's just equivalent to startIndex == 1.

File: master/internal/api.go (hunk 1 above)
Comment (hamidzr) [reply master/internal/api.go:50]: scratch that example. endIndex is exclusive but startIndex is inclusive.

**Comment by hamidzr** [comment]
I worked on adding some tests which resulted in some necessary refactoring as well. I'll put up those as a separate PR so this fix can go in sooner rather than later 

https://github.com/determined-ai/determined/pull/1504

**Pull request merged**